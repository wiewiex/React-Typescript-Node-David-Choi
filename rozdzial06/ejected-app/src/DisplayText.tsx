import React, { useState, FC } from "react";
import UserTodos from "./UserTodos";

interface DisplayTextProps {
    getUserFullName: (username: string) => Promise<string>;
}

const DisplayText: FC<DisplayTextProps> = ({ getUserFullName }) => {
    const [txt, setTxt] = useState("");
    const [msg, setMsg] = useState("");
    const [todoControl, setTodoControl] = useState<ReturnType<typeof UserTodos>>();

    const onChangeTxt = (e:React.ChangeEvent<HTMLInputElement>) => {
        setTxt(e.target.value);
    }    

    const onClickShowMsg = async (e: React.MouseEvent<HTMLButtonElement, MouseEvent>) => {
        e.preventDefault();    
        setMsg(`Witaj na zajÄ™ciach z testowania Reacta, ${await getUserFullName(txt)}!`);
        setTodoControl(<UserTodos username={txt} />);        
    }    

    return (
        <form>
            <div>
                <label>Jak masz na imie?</label>
            </div>
            <div>
                <input data-testid="user-input"
                    value={txt} onChange={onChangeTxt} />
            </div>
            <div>
                <button data-testid="input-submit"
                    onClick={onClickShowMsg}>Wyswietl komunikat</button>
            </div>
            <div>
                <label data-testid="final-msg">{msg}</label>
            </div>
            {todoControl}            
        </form>
    )
}

export default DisplayText;