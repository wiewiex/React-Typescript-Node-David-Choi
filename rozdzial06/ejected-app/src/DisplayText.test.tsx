import { render, fireEvent, cleanup, waitFor } from "@testing-library/react";
import DisplayText from "./DisplayText";
import "@testing-library/jest-dom/extend-expect";

jest.mock("./UserTodos");

describe("Test komponentu DisplayText", () => {
    const userFullName = "Janek Tester";

    const getUserFullNameMock = (
        username: string
    ) : [Promise<string>, jest.Mock<Promise<string>, [string]>] => {
        const promise = new Promise<string>((res, rej) => {
            res(userFullName);
        });
        const getUserFullName = jest.fn(
            async (username: string): Promise<string> => {
                return promise;
            }
        );

        return [promise, getUserFullName];
    };


    it("jest wyswietlany bez problemow", () => {
        const username= "testuser";
        const [promise, getUserFullName] = getUserFullNameMock(username);
        const { baseElement } = render(<DisplayText getUserFullName={getUserFullName} />);      
        expect(baseElement).toBeInTheDocument();
    });

    it("jest zgodny z migawką", () => {
        const username= "testuser";
        const [promise, getUserFullName] = getUserFullNameMock(username);
        const { baseElement } = render(<DisplayText getUserFullName={getUserFullName} />);  
        expect(baseElement).toMatchSnapshot();
    })

    it("uzyskuje text z pola", () => {
        const username = "testuser";
        const [promise, getUserFullName] = getUserFullNameMock(username);
        const { getByTestId } = render(<DisplayText getUserFullName={getUserFullName} />);
        const input = getByTestId("user-input");
        fireEvent.change(input, { target: { value: username }});
        expect(input).toBeInTheDocument();
        expect(input).toHaveValue(username);
    });

    // it("wyswietla komunikat powitalny", async () => {
    //     const username = "testuser";
    //     const [promise, getUserFullName] = getUserFullNameMock(userFullName);

    //     const msg = `Witaj na zajęciach z testowania Reacta, ${username}!`;
    //     const { getByTestId } = render(<DisplayText getUserFullName={getUserFullName} />);
    //     const input = getByTestId("user-input");
    //     const label = getByTestId("final-msg");
    //     fireEvent.change(input, { target: { value: username } });
    //     const btn = getByTestId("input-submit");
    //     fireEvent.click(btn);

    //     expect(label).toBeInTheDocument();
    //     await waitFor(() => promise);
    //     expect(label.innerHTML).toBe(msg);
    // })
})

