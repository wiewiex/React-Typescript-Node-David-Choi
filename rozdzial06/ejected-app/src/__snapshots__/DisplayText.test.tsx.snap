// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test komponentu DisplayText jest zgodny z migawkÄ… 1`] = `
<body>
  <div>
    <form>
      <div>
        <label>
          Jak masz na imie?
        </label>
      </div>
      <div>
        <input
          data-testid="user-input"
          value=""
        />
      </div>
      <div>
        <button
          data-testid="input-submit"
        >
          Wyswietl komunikat
        </button>
      </div>
      <div>
        <label
          data-testid="final-msg"
        />
      </div>
    </form>
  </div>
</body>
`;
